Drop gif + png, and redirect png to raster.shields.io

paulmelnikow opened the pull request:
This PR does a few things:

1. Remove rasterization support from the server. This responsibility is delegated to a raster server which proxies the SVG badges and renders them.
2. When a raster server URL is configured, 301 redirect all .png badges to the identical URL on the raster server.
    `https://img.shields.io/npm/v/express.png?style=flat-square` ↪️`https://raster.shields.io/npm/v/express.png?style=flat-square`
3. For configured redirects, redirect to the canonical URL on the raster server.
    `https://img.shields.io/vso/build/totodem/8cf3ec0e-d0c2-4fcd-8206-ad204f254a96/2.png?style=flat-square`
    ↪️`https://img.shields.io/azure-devops/build/totodem/8cf3ec0e-d0c2-4fcd-8206-ad204f254a96/2.png?style=flat-square`
4. Redirect the "legacy badge old version" to the appropriate URL on the raster server.
5. When no raster server is configured (e.g. PRs), render an SVG containing **404 | raster badges not available** for all `.png` badges. (Note that the raster server can be self-hosted; however, this is deferred to a later PR.)
5. Drop support for jpg and gif which are very infrequently used (see #3112). Render an SVG containing **410 | jpg no longer available**.
7. ~~Remove raster dependencies.~~ Remove the raster cache (which is only used in the CLI, and therefore pointless).
8. Move the LRUCache code out of the npm package.
8. A wee bit of refactoring in `server.js`.

Ref #3112
Close #3631

chris48s commented:
I've not read over all this, but giving it a cursory look, it seems like you're completely removing raster support from the package/CLI in this PR whereas a couple of weeks ago I think you were looking at leaving support in but converting `gm` to a peer dependency: https://github.com/badges/shields/issues/3587#issuecomment-504014841

Is that an oversight or has your opinion on raster support changed since we established that its only the CLI that supports them and not the library: https://github.com/badges/shields/pull/3586#discussion_r295802787 ?

paulmelnikow commented:
> I've not read over all this, but giving it a cursory look, it seems like you're completely removing raster support from the package/CLI in this PR whereas a couple of weeks ago I think you were looking at leaving support in but converting `gm` to a peer dependency: [#3587 (comment)](https://github.com/badges/shields/issues/3587#issuecomment-504014841)
> 
> Is that an oversight or has your opinion on raster support changed since we established that its only the CLI that supports them and not the library: [#3586 (comment)](https://github.com/badges/shields/pull/3586#discussion_r295802787) ?

Ah, right, I forgot about that. Thanks for the reminder! 😝

I'm fine with restoring that code it and keeping it as a peer dep, though another alternative would be to swap in the more modern rasterization lib [convert-svg-to-png](https://www.npmjs.com/package/convert-svg-to-png) that's being used in svg-to-image-proxy.

This also touches lru-cache. I don't want to smush your work on v3; not really sure the best way forward with these changes at the boundary of the library.

To keep this smallish, what do you think about my adding the raster capability back, using the new dependency, in a separate PR?

paulmelnikow commented:
Come to think of it, it wouldn't be so hard to put the raster support back into `gh-badges`. I could revert the removal of the `gm` dependency and the `lru-cache` changes. Pretty much all the other removals (e.g. references to imagemagick in the tutorial) could stay as they are.

chris48s commented:
I don't think the change to the LRU cache makes any difference because it doesn't affect the public interface (or at least, the documented interface) of the package - its just internals.

The bits that change the documented functionality (i.e: the CLI) are best done in a separate PR targetted at the `v3-dev` branch (same as I've done with #3586 ).

paulmelnikow commented:
I just remembered that raster badges are broken in the CLI, which makes this sort of moot, though I reverted it for now. I kept the lru-cache in its new location and removed internal references to it instead, as even if the CLI worked, the cache would be pointless.

chris48s commented:
> I just remembered that raster badges are broken in the CLI

Are they? If I run `badge build passed :green .png > mybadge.png` it spits out:

![mybadge](https://user-images.githubusercontent.com/6025893/60760260-5b291380-a02a-11e9-8380-5902de337e72.png)

paulmelnikow commented:
Huh, after upgrading imagemagick, this is what I get, which is obviously wrong, though different from what I was seeing when I opened #1788.

![mybadge](https://user-images.githubusercontent.com/1487036/60760350-2c974280-a002-11e9-9bbf-12f5e729a427.png)

chris48s commented on config/custom-environment-variables.yml:
```diff
@@ -14,6 +14,8 @@ public:
 
   redirectUri: 'REDIRECT_URI'
 
+  rasterUrl: 'RASTER_URL'
```
Do you think it makes sense to configure a `RASTER_URL` in staging so that something like https://shields-staging-pr-3644.herokuapp.com/appveyor/ci/gruntjs/grunt.png works, or does it just become confusing to mix staging and production architecture?

paulmelnikow replied:
Yea, I was thinking it would be nice to have one for the staging environment, though perhaps it's not necessary for PRs. I _think_ we could configure the env var in the staging app and that way it'll apply only to staging (not PRs).

chris48s replied:
its something we can deal with as an env var in heroku anyway, rather than something that needs changing in this PR

chris48s commented on core/base-service/base.spec.js:
```diff
@@ -316,7 +316,7 @@ describe('BaseService', function() {
 
   describe('ScoutCamp integration', function() {
-    const expectedRouteRegex = /^\/foo\/([^/]+?)\.(svg|png|gif|jpg|json)$/
+    const expectedRouteRegex = /^\/foo\/([^/]+?)\.(svg|json)$/
```
While we're fiddling with this, is this also a sensible point in time to introduce extensionless SVG https://github.com/badges/shields/issues/2674 or not?

paulmelnikow replied:
Yea, it's a good point. I was thinking about that, but I'm eager to deploy this so we can find out how the raster server is performing. Extensionless SVG will need a few other changes made including some CDN config tweaks. I'm curious to find out how stable the raster server is and don't really want to hold this up.

chris48s replied:
Ok. Lets stick to changing one thing at a time :+1:

chris48s commented on doc/self-hosting.md:
```diff
@@ -93,6 +93,23 @@ machine.
+  for the legacy raster URLs instead of 404's.
+
+If anyone has set this up, more documentation on how to do this would be
+welcome! It would also be nice to ship a Docker image that includes a
```
Do you think its worth adding a sentence just saying its designed to run as a lambda on now/AWS/etc, or is that stuff just best covered in the README for `svg-to-image-proxy`?

paulmelnikow replied:
I think that's a fine thing to add. If it's information that isn't going to change, it's not the worst thing to mention it in two places.

chris48s commented on Dockerfile:
```diff
@@ -1,7 +1,5 @@
 FROM node:8-alpine
 
-RUN apk add --no-cache gettext imagemagick librsvg git
```
do we definitely not need git anymore?

paulmelnikow replied:
It looks like the automated Docker build ran successfully on this branch, and I can't think of a reason why it would be needed at runtime.

<img width="1100" alt="Screen Shot 2019-07-06 at 3 47 19 PM" src="https://user-images.githubusercontent.com/1487036/60760565-5dc54200-a005-11e9-9d53-e64a01b731c3.png">

Link: https://cloud.docker.com/u/shieldsio/repository/registry-1.docker.io/shieldsio/shields/timeline

paulmelnikow commented:
Gosh, shall we ship this? 🚀

paulmelnikow commented:
1. Working: https://shields-raster-staging.now.sh/badge/foo-bar-blue.png?style=flat-square

2. Not redirecting correctly: http://shields-staging.herokuapp.com/badge/foo-bar-blue.png?style=flat-square
    It's dropping `?style=flat-square` and redirecting to a nonsense static badge:
    https://shields-raster-staging.now.sh/badge/badge-foo--bar--blue-lightgray.png

3. Also not redirecting correctly: http://shields-staging.herokuapp.com/npm/v/express.png?style=flat-square

paulmelnikow commented:
Confirmed fixed in #3659. (If it's not fixed for you, the bad redirects are probably stuck in your cache.)
